File: services/pages.go
package services

import (
	"context"
	"errors"
	"fmt"
	"path"
	"sort"
	"strings"
	"unicode/utf8"

	"github.com/fmind/fkf/core"
)

// wiki/ and projects/ are the same layer with different required frontmatter, so they share
// the parser, the listing, and the validator. A flat layer has no directories to navigate by,
// which is why the tag index is the navigation surface rather than a convenience.

// PageListing is what `fkf list wiki` and `fkf list projects` return.
type PageListing struct {
	Layer core.Layer `json:"layer"`
	Pages []Page     `json:"pages"`
	Total int        `json:"total"`
}

// PageFilter narrows a listing.
type PageFilter struct {
	Tags   []string
	Status string
	Type   string
	Limit  int
}

// ListPages returns the pages of a flat Markdown layer, filtered and in slug order.
func ListPages(ctx context.Context, base *Base, layer core.Layer, filter PageFilter) (*PageListing, error) {
	pages, _, err := loadMarkdownLayer(ctx, base, layer)
	if err != nil {
		return nil, err
	}
	// `--status wibble` used to return "0 pages", exit 0 — silently indistinguishable from
	// zero paused projects. Status is a closed three-value set regardless of what is in this
	// base, so it is checked before the pages a typo would have hidden from the caller.
	if filter.Status != "" {
		if err := requireKnown("status", []string{filter.Status}, ProjectStatuses); err != nil {
			return nil, err
		}
	}
	// Tags are this base's own vocabulary rather than a fixed set, so they are checked against
	// what these very pages already carry — no second scan. The check folds case the same way
	// hasTag's actual match does below, or a validly-cased filter value could be refused as
	// unknown right before the match that would have accepted it.
	if len(filter.Tags) > 0 {
		known, display := map[string]bool{}, presentTags(pages)
		for _, tag := range display {
			known[strings.ToLower(tag)] = true
		}
		for _, wanted := range filter.Tags {
			if !known[strings.ToLower(strings.TrimSpace(wanted))] {
				return nil, unknownValueError("tag", wanted, display)
			}
		}
	}
	listing := &PageListing{Layer: layer, Pages: []Page{}}
	for _, page := range pages {
		if err := checkContext(ctx); err != nil {
			return nil, err
		}
		if !matchesFilter(page, filter) {
			continue
		}
		page.Body, page.Links, page.Headings = "", nil, nil
		listing.Pages = append(listing.Pages, page)
	}
	listing.Total = len(listing.Pages)
	if filter.Limit > 0 && len(listing.Pages) > filter.Limit {
		listing.Pages = listing.Pages[:filter.Limit]
	}
	return listing, nil
}

// matchesFilter requires every requested tag rather than any of them. Tags narrow a flat
// namespace, and an any-match filter over a real vocabulary returns almost everything.
func matchesFilter(page Page, filter PageFilter) bool {
	if filter.Status != "" && page.Status != filter.Status {
		return false
	}
	if filter.Type != "" && page.Type != filter.Type {
		return false
	}
	for _, wanted := range filter.Tags {
		if !hasTag(page.Tags, wanted) {
			return false
		}
	}
	return true
}

// presentTags is the sorted, deduped set of tags these already-loaded pages carry, in the
// casing they were first authored with — display form for the refusal message, not the fold
// used to compare against it.
func presentTags(pages []Page) []string {
	seen := map[string]string{}
	for _, page := range pages {
		for _, tag := range page.Tags {
			key := strings.ToLower(strings.TrimSpace(tag))
			if _, exists := seen[key]; !exists {
				seen[key] = tag
			}
		}
	}
	tags := make([]string, 0, len(seen))
	for _, display := range seen {
		tags = append(tags, display)
	}
	sort.Strings(tags)
	return tags
}

func hasTag(tags []string, wanted string) bool {
	wanted = strings.ToLower(strings.TrimSpace(wanted))
	for _, tag := range tags {
		if strings.ToLower(tag) == wanted {
			return true
		}
	}
	return false
}

// ReadPageBySlug returns one page of a flat layer.
func ReadPageBySlug(base *Base, layer core.Layer, slug string) (Page, error) {
	return ReadPageBySlugContext(context.Background(), base, layer, slug)
}

// ReadPageBySlugContext returns one page of a flat layer with cooperative cancellation.
func ReadPageBySlugContext(ctx context.Context, base *Base, layer core.Layer, slug string) (Page, error) {
	slug = strings.TrimSuffix(strings.TrimSpace(slug), core.MarkdownExtension)
	if slug == "" {
		return Page{}, fmt.Errorf("a %s page is addressed by its slug, for example retrieval-boundary", layer)
	}
	return ReadPageContext(ctx, base, path.Join(string(layer), slug+core.MarkdownExtension))
}

// TagCount is one tag and the pages carrying it.
type TagCount struct {
	Tag   string   `json:"tag"`
	Count int      `json:"count"`
	Pages []string `json:"pages"`
}

// TagVocabulary is a layer's complete tag vocabulary with its usage.
type TagVocabulary struct {
	Layer    core.Layer `json:"layer"`
	Tags     []TagCount `json:"tags"`
	Untagged []string   `json:"untagged,omitempty"`
	Pages    int        `json:"pages"`
}

// BuildTagVocabulary groups a layer's pages by tag, most-used first. The histogram is what makes a
// vocabulary legible, and reusing it is what stops a wiki growing four spellings of one idea.
func BuildTagVocabulary(ctx context.Context, base *Base, layer core.Layer) (*TagVocabulary, error) {
	pages, _, err := loadMarkdownLayer(ctx, base, layer)
	if err != nil {
		return nil, err
	}
	index := &TagVocabulary{Layer: layer, Tags: []TagCount{}, Pages: len(pages)}
	grouped := map[string][]string{}
	for _, page := range pages {
		if err := checkContext(ctx); err != nil {
			return nil, err
		}
		if len(page.Tags) == 0 {
			index.Untagged = append(index.Untagged, page.Slug)
			continue
		}
		for _, tag := range page.Tags {
			normalized := strings.ToLower(strings.TrimSpace(tag))
			grouped[normalized] = append(grouped[normalized], page.Slug)
		}
	}
	for tag, slugs := range grouped {
		sort.Strings(slugs)
		index.Tags = append(index.Tags, TagCount{Tag: tag, Count: len(slugs), Pages: slugs})
	}
	sort.Slice(index.Tags, func(i, j int) bool {
		if index.Tags[i].Count == index.Tags[j].Count {
			return index.Tags[i].Tag < index.Tags[j].Tag
		}
		return index.Tags[i].Count > index.Tags[j].Count
	})
	sort.Strings(index.Untagged)
	return index, nil
}

// SearchHit is one Markdown document matching a search, with the excerpt that explains why.
// Layer and Date are carried on the hit rather than only on the enclosing result because
// `fkf find` returns hits from several layers at once and a reader has to tell them apart.
type SearchHit struct {
	URI     string     `json:"uri"`
	Layer   core.Layer `json:"layer,omitempty"`
	Slug    string     `json:"slug"`
	Title   string     `json:"title,omitempty"`
	Type    string     `json:"type,omitempty"`
	Date    string     `json:"date,omitempty"`
	Tags    []string   `json:"tags,omitempty"`
	Score   int        `json:"score"`
	Matched []string   `json:"matched"`
	Excerpt string     `json:"excerpt,omitempty"`
}

// SearchResult is the layer-scoped lexical result used by the universal `fkf find` command.
type SearchResult struct {
	Layer core.Layer  `json:"layer"`
	Terms []string    `json:"terms"`
	Hits  []SearchHit `json:"hits"`
}

// SearchPages is a lexical, deterministic scan: a title or tag match outweighs a body match,
// and every hit reports which terms matched. There is no ranking model and no index engine —
// a flat layer of a few hundred pages is a scan, and a scan is explainable.
func SearchPages(ctx context.Context, base *Base, layer core.Layer, terms []string, filter PageFilter) (*SearchResult, error) {
	pages, _, err := loadMarkdownLayer(ctx, base, layer)
	if err != nil {
		return nil, err
	}
	normalized := normalizeTerms(terms)
	if len(normalized) == 0 {
		return nil, errors.New("search needs at least one term")
	}
	result := &SearchResult{Layer: layer, Terms: normalized, Hits: []SearchHit{}}
	for _, page := range pages {
		if err := checkContext(ctx); err != nil {
			return nil, err
		}
		if !matchesFilter(page, filter) {
			continue
		}
		if hit, ok := scorePage(page, normalized); ok {
			hit.Layer = layer
			result.Hits = append(result.Hits, hit)
		}
	}
	SortSearchHits(result.Hits)
	if filter.Limit > 0 && len(result.Hits) > filter.Limit {
		result.Hits = result.Hits[:filter.Limit]
	}
	return result, nil
}

// normalizeTerms lowercases and drops the empties once, so every lexical caller compares the
// same way. `fkf find` assembles its terms from four flags and would otherwise repeat this.
func normalizeTerms(terms []string) []string {
	normalized := make([]string, 0, len(terms))
	for _, term := range terms {
		if trimmed := strings.ToLower(strings.TrimSpace(term)); trimmed != "" {
			normalized = append(normalized, trimmed)
		}
	}
	return normalized
}

func scorePage(page Page, terms []string) (SearchHit, bool) {
	haystackTitle := strings.ToLower(page.Title + " " + page.Slug + " " + page.Description)
	haystackTags := strings.ToLower(strings.Join(page.Tags, " "))
	haystackBody := strings.ToLower(page.Body)
	hit := SearchHit{
		URI: page.URI, Slug: page.Slug, Title: page.Title,
		Type: page.Type, Tags: page.Tags, Matched: []string{},
	}
	for _, term := range terms {
		points := 0
		switch {
		case strings.Contains(haystackTitle, term):
			points = 10
		case strings.Contains(haystackTags, term):
			points = 6
		case strings.Contains(haystackBody, term):
			points = 2
		}
		if points == 0 {
			// Grep terms are conjunctive across records and pages. Returning a partial page
			// match here made one command silently switch from AND to OR by layer.
			return SearchHit{}, false
		}
		hit.Score += points
		hit.Matched = append(hit.Matched, term)
		if hit.Excerpt == "" {
			hit.Excerpt = excerptAround(page.Body, term)
		}
	}
	return hit, true
}

// excerptAround returns a bounded window of the body around the first match, so a hit can be
// judged without opening the page.
func excerptAround(body, term string) string {
	foldedBody := strings.ToLower(body)
	byteIndex := strings.Index(foldedBody, term)
	if byteIndex < 0 {
		return ""
	}
	// Unicode lowercasing preserves rune positions, but not byte positions. Slice the original
	// text by rune index so a case-folded match cannot produce invalid or reversed byte bounds.
	bodyRunes := []rune(body)
	matchIndex := utf8.RuneCountInString(foldedBody[:byteIndex])
	matchLength := utf8.RuneCountInString(term)
	const radius = 90
	start, end := max(0, matchIndex-radius), min(len(bodyRunes), matchIndex+matchLength+radius)
	excerpt := strings.Join(strings.Fields(string(bodyRunes[start:end])), " ")
	if start > 0 {
		excerpt = "…" + excerpt
	}
	if end < len(bodyRunes) {
		excerpt += "…"
	}
	return excerpt
}

File: src/fkf/pages.py
"""Flat authored-page listing, vocabulary, and deterministic lexical search."""

from __future__ import annotations

import os
from dataclasses import dataclass, field, replace
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Final

from fkf.base import Base
from fkf.config import ConfigError
from fkf.io import read_file_limited
from fkf.markdown import PROJECT_STATUSES, Page, parse_page
from fkf.process import Cancellation, check_cancel
from fkf.store import MARKDOWN_EXTENSION, MAX_NARRATIVE_BYTES, Layer

if TYPE_CHECKING:
    from fkf.scan import ScanGuard

MAX_VOCABULARY_LISTED: Final = 20


@dataclass(frozen=True, slots=True)
class PageFilter:
    tags: tuple[str, ...] = ()
    status: str = ""
    type: str = ""
    limit: int = 0


@dataclass(frozen=True, slots=True)
class PageListing:
    layer: Layer
    pages: tuple[Page, ...]
    total: int


@dataclass(frozen=True, slots=True)
class TagCount:
    tag: str
    count: int
    pages: tuple[str, ...]


@dataclass(frozen=True, slots=True)
class TagVocabulary:
    layer: Layer
    tags: tuple[TagCount, ...]
    untagged: tuple[str, ...] = field(metadata={"json": "untagged,omitempty"})
    pages: int


@dataclass(frozen=True, slots=True)
class SearchHit:
    uri: str
    layer: Layer
    slug: str
    title: str = ""
    type: str = ""
    date: str = ""
    tags: tuple[str, ...] = ()
    score: int = 0
    matched: tuple[str, ...] = ()
    excerpt: str = ""


@dataclass(frozen=True, slots=True)
class SearchResult:
    layer: Layer
    terms: tuple[str, ...]
    hits: tuple[SearchHit, ...]


def _join_vocabulary(vocabulary: tuple[str, ...]) -> str:
    if not vocabulary:
        return "none"
    if len(vocabulary) <= MAX_VOCABULARY_LISTED:
        return ", ".join(vocabulary)
    shown = vocabulary[:MAX_VOCABULARY_LISTED]
    return f"{', '.join(shown)}, … and {len(vocabulary) - MAX_VOCABULARY_LISTED} more (see `fkf config`)"


def require_known(field: str, values: tuple[str, ...], vocabulary: tuple[str, ...]) -> None:
    """Reject typos in a closed selector before they become authoritative empty results."""
    known = set(vocabulary)
    for value in values:
        if value not in known:
            raise ConfigError(f"unknown {field} {value!r}; this base declares {_join_vocabulary(vocabulary)}")


def read_page(
    base: Base,
    uri: str,
    *,
    cancel: Cancellation | None = None,
    scan: ScanGuard | None = None,
) -> Page:
    check_cancel(cancel)
    absolute = base.store.resolve(uri)
    data = read_file_limited(absolute, MAX_NARRATIVE_BYTES)
    if scan is not None:
        # Charge the inode bytes actually opened, not racy path metadata.
        scan.consume(len(data))
    check_cancel(cancel)
    try:
        modified = datetime.fromtimestamp(absolute.stat().st_mtime, tz=UTC)
    except OSError:
        modified = None
    page = parse_page(uri, data, modified)
    check_cancel(cancel)
    return page


def load_markdown_layer(
    base: Base,
    layer: Layer,
    *,
    cancel: Cancellation | None = None,
    scan: ScanGuard | None = None,
    metadata_only: bool = False,
) -> tuple[tuple[Page, ...], tuple[str, ...]]:
    check_cancel(cancel)
    directory = base.store.directory(layer)
    try:
        iterator = os.scandir(directory)
    except FileNotFoundError:
        return (), ()
    except OSError as error:
        raise OSError(f"list {directory}: {error}") from error
    pages: list[Page] = []
    nested: list[str] = []
    with iterator:
        entries = []
        for entry in iterator:
            check_cancel(cancel)
            if scan is not None:
                scan.visit()
            entries.append(entry)
        for entry in sorted(entries, key=lambda item: item.name):
            check_cancel(cancel)
            if entry.is_dir(follow_symlinks=False):
                nested.append(f"{layer}/{entry.name}/")
                continue
            if not entry.name.endswith(MARKDOWN_EXTENSION):
                continue
            uri = f"{layer}/{entry.name}"
            page = read_page(base, uri, cancel=cancel, scan=scan)
            if metadata_only:
                page = replace(page, body="", links=(), headings=())
            pages.append(page)
    return tuple(sorted(pages, key=lambda page: page.uri)), tuple(nested)


def _present_tags(pages: tuple[Page, ...], cancel: Cancellation | None) -> tuple[str, ...]:
    seen: dict[str, str] = {}
    for page in pages:
        check_cancel(cancel)
        for tag in page.tags:
            seen.setdefault(tag.strip().lower(), tag)
    return tuple(sorted(seen.values()))


def _has_tag(tags: tuple[str, ...], wanted: str) -> bool:
    normalized = wanted.strip().lower()
    return any(tag.lower() == normalized for tag in tags)


def _matches_filter(page: Page, filters: PageFilter) -> bool:
    if filters.status and page.status != filters.status:
        return False
    if filters.type and page.type != filters.type:
        return False
    return all(_has_tag(page.tags, wanted) for wanted in filters.tags)


def _validate_filter(pages: tuple[Page, ...], filters: PageFilter, cancel: Cancellation | None) -> None:
    if filters.limit < 0:
        raise ValueError("limit must not be negative")
    if filters.status:
        require_known("status", (filters.status,), PROJECT_STATUSES)
    if filters.tags:
        display = _present_tags(pages, cancel)
        known = {tag.strip().lower() for tag in display}
        for wanted in filters.tags:
            if wanted.strip().lower() not in known:
                raise ConfigError(f"unknown tag {wanted!r}; this base declares {_join_vocabulary(display)}")


def list_pages(
    base: Base,
    layer: Layer,
    filters: PageFilter | None = None,
    *,
    cancel: Cancellation | None = None,
    scan: ScanGuard | None = None,
    metadata_only: bool = False,
) -> PageListing:
    filters = filters or PageFilter()
    pages, _nested = load_markdown_layer(
        base,
        layer,
        cancel=cancel,
        scan=scan,
        metadata_only=metadata_only,
    )
    _validate_filter(pages, filters, cancel)
    selected_pages: list[Page] = []
    for page in pages:
        check_cancel(cancel)
        if _matches_filter(page, filters):
            selected = replace(page, body="", links=(), headings=())
            if scan is not None:
                scan.retain(selected)
            selected_pages.append(selected)
    selected = tuple(selected_pages)
    total = len(selected)
    if filters.limit:
        selected = selected[: filters.limit]
    return PageListing(layer, selected, total)


def build_tag_vocabulary(
    base: Base,
    layer: Layer,
    *,
    cancel: Cancellation | None = None,
    scan: ScanGuard | None = None,
    metadata_only: bool = False,
) -> TagVocabulary:
    pages, _nested = load_markdown_layer(
        base,
        layer,
        cancel=cancel,
        scan=scan,
        metadata_only=metadata_only,
    )
    grouped: dict[str, list[str]] = {}
    untagged: list[str] = []
    for page in pages:
        check_cancel(cancel)
        if not page.tags:
            untagged.append(page.slug)
            continue
        for tag in page.tags:
            grouped.setdefault(tag.strip().lower(), []).append(page.slug)
    counts = [TagCount(tag, len(slugs), tuple(sorted(slugs))) for tag, slugs in grouped.items()]
    counts.sort(key=lambda item: (-item.count, item.tag))
    sorted_untagged = tuple(sorted(untagged))
    if scan is not None:
        # The resource returns tags and untagged slugs, not its scanned input pages.
        for item in counts:
            scan.retain(item)
        for slug in sorted_untagged:
            scan.retain(slug)
    return TagVocabulary(layer, tuple(counts), sorted_untagged, len(pages))


def normalize_terms(terms: tuple[str, ...] | list[str]) -> tuple[str, ...]:
    return tuple(term.strip().lower() for term in terms if term.strip())


def _excerpt_around(body: str, term: str) -> str:
    folded = body.lower()
    index = folded.find(term)
    if index < 0:
        return ""
    radius = 90
    start = max(0, index - radius)
    end = min(len(body), index + len(term) + radius)
    excerpt = " ".join(body[start:end].split())
    if start:
        excerpt = f"…{excerpt}"
    if end < len(body):
        excerpt = f"{excerpt}…"
    return excerpt


def _score_page(page: Page, layer: Layer, terms: tuple[str, ...]) -> SearchHit | None:
    title = " ".join((page.title, page.slug, page.description, *page.aliases)).lower()
    tags = " ".join(page.tags).lower()
    body = page.body.lower()
    score = 0
    excerpt = ""
    for term in terms:
        if term in title:
            points = 10
        elif term in tags:
            points = 6
        elif term in body:
            points = 2
        else:
            return None
        score += points
        if not excerpt:
            excerpt = _excerpt_around(page.body, term)
    return SearchHit(
        uri=page.uri,
        layer=layer,
        slug=page.slug,
        title=page.title,
        type=page.type,
        tags=page.tags,
        score=score,
        matched=terms,
        excerpt=excerpt,
    )


def search_pages(
    base: Base,
    layer: Layer,
    terms: tuple[str, ...] | list[str],
    filters: PageFilter | None = None,
    *,
    cancel: Cancellation | None = None,
) -> SearchResult:
    filters = filters or PageFilter()
    pages, _nested = load_markdown_layer(base, layer, cancel=cancel)
    _validate_filter(pages, filters, cancel)
    normalized = normalize_terms(terms)
    if not normalized:
        raise ValueError("search needs at least one term")
    hits = []
    for page in pages:
        check_cancel(cancel)
        if _matches_filter(page, filters) and (hit := _score_page(page, layer, normalized)):
            hits.append(hit)
    hits.sort(key=lambda hit: (-hit.score, hit.uri))
    if filters.limit:
        hits = hits[: filters.limit]
    return SearchResult(layer, normalized, tuple(hits))


__all__ = [
    "MAX_VOCABULARY_LISTED",
    "PageFilter",
    "PageListing",
    "SearchHit",
    "SearchResult",
    "TagCount",
    "TagVocabulary",
    "build_tag_vocabulary",
    "list_pages",
    "load_markdown_layer",
    "normalize_terms",
    "read_page",
    "require_known",
    "search_pages",
]
